src/doc/trpl/closures.md

```diff
@@ -1,214 +1,513 @@
-concurrency features, and so we'll just leave it at this for
-now. We'll talk about them more in the "Concurrency" chapter of the book.
+# Closure implementation
 
-## Accepting closures as arguments
+Rust's implementation of closures is a bit different than other languages. They
+are effectively syntax sugar for traits. You'll want to make sure to have read
+the [traits chapter][traits] before this one, as well as the chapter on [static
+and dynamic dispatch][dispatch], which talks about trait objects.
 
-Closures are most useful as an argument to another function. Here's an example:
+[traits]: traits.html
+[dispatch]: static-and-dynamic-dispatch.html
 
-```{rust}
-fn twice<F: Fn(i32) -> i32>(x: i32, f: F) -> i32 {
-    f(x) + f(x)
+Got all that? Good.
+
+The key to understanding how closures work under the hood is something a bit
+strange: `()` is an overloadable operator. From this, everything else clicks
+into place. In Rust, we use the trait system to overload operators. `()` is no
+different. We have three separate traits to overload `()` with:
+
+```rust
+# mod foo {
+pub trait Fn<Args> {
+    type Output;
+
+    extern "rust-call" fn call(&self, args: Args) -> Self::Output;
 }
 
-fn main() {
-    let square = |x: i32| { x * x };
+pub trait FnMut<Args> {
+    type Output;
 
-    twice(5, square); // evaluates to 50
+    extern "rust-call" fn call_mut(&mut self, args: Args) -> Self::Output;
 }
+
+pub trait FnOnce<Args> {
+    type Output;
+
+    extern "rust-call" fn call_once(self, args: Args) -> Self::Output;
+}
+# }
 ```
 
-Let's break the example down, starting with `main`:
+You'll notice that the primary difference between the three is `self`: `Fn`
+takes `&self`, `FnMut` takes `&mut self`, and `FnOnce` takes `self`. This
+covers all three kinds of `self` via the usual method call syntax. But we've
+split them up into three traits, rather than having a single one. This gives
+us a large amount of control over what kind of closures we can take.
+
+The `|| {}` syntax for closures is sugar for these three traits. For example,
+when you create a closure like this:
 
-```{rust}
-let square = |x: i32| { x * x };
+```rust
+let num = 5;
+let add_num = |x| x + num;
+
+add_num(1);
 ```
 
-We've seen this before. We make a closure that takes an integer, and returns
-its square.
+Rust desugars that code to this code:
 
-```{rust}
-# fn twice<F: Fn(i32) -> i32>(x: i32, f: F) -> i32 { f(x) + f(x) }
-# let square = |x: i32| { x * x };
-twice(5, square); // evaluates to 50
 ```
+# #![feature(unboxed_closures)]
 
-This line is more interesting. Here, we call our function, `twice`, and we pass
-it two arguments: an integer, `5`, and our closure, `square`. This is just like
-passing any other two variable bindings to a function, but if you've never
-worked with closures before, it can seem a little complex. Just think: "I'm
-passing two variables: one is an i32, and one is a function."
+struct ClosureEnvironment<'env> {
```

Comment:
I'd remove `Environment` from this name. Also, I've basically only ever seen non-single-letter lifetimes in rustc's internals, so `'a` might be more familiar.